File: 1-sliding-window-log-vs-counter/1-java/src/main/java/com/starci/slidingwindow/SlidingWindowController.java
package com.starci.slidingwindow;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.LinkedHashMap;
import java.util.Map;

@RestController
@RequestMapping("/api/sliding-window")
public class SlidingWindowController {

    private final SlidingWindowService service;

    public SlidingWindowController(SlidingWindowService service) {
        this.service = service;
    }

    public record CheckRequest(String key, long limit, long windowMs) {
    }

    @PostMapping("/log/check")
    public Map<String, Object> logCheck(@RequestBody CheckRequest req) {
        SlidingWindowService.LimitResult r = service.checkLog(req.key(), req.limit(), req.windowMs());
        return body("log", r, req.limit());
    }

    @PostMapping("/counter/check")
    public Map<String, Object> counterCheck(@RequestBody CheckRequest req) {
        SlidingWindowService.LimitResult r = service.checkCounter(req.key(), req.limit(), req.windowMs());
        return body("counter", r, req.limit());
    }

    @GetMapping("/compare")
    public SlidingWindowService.CompareResult compare(
            @RequestParam(defaultValue = "demo") String key,
            @RequestParam(defaultValue = "5") long limit,
            @RequestParam(defaultValue = "2000") long windowMs,
            @RequestParam(defaultValue = "10") int n) {
        return service.compare(key, limit, windowMs, n);
    }

    // Shared contract: { algorithm, allowed, count, limit }.
    private Map<String, Object> body(String algorithm, SlidingWindowService.LimitResult r, long limit) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("algorithm", algorithm);
        out.put("allowed", r.allowed());
        out.put("count", r.count());
        out.put("limit", limit);
        return out;
    }
}

File: 1-sliding-window-log-vs-counter/0-typescript/src/sliding-window.controller.ts
import { Body, Controller, Get, Post, Query } from "@nestjs/common"
import { SlidingWindowService } from "./sliding-window.service"

interface CheckBody {
    key: string
    limit: number
    windowMs: number
}

@Controller("api/sliding-window")
export class SlidingWindowController {
    public constructor(private readonly service: SlidingWindowService) {}

    @Post("log/check")
    public async logCheck(@Body() body: CheckBody): Promise<Record<string, unknown>> {
        const r = await this.service.checkLog(body.key, body.limit, body.windowMs)
        // Shared contract: { algorithm, allowed, count, limit }.
        return { algorithm: "log", allowed: r.allowed, count: r.count, limit: body.limit }
    }

    @Post("counter/check")
    public async counterCheck(@Body() body: CheckBody): Promise<Record<string, unknown>> {
        const r = await this.service.checkCounter(body.key, body.limit, body.windowMs)
        return { algorithm: "counter", allowed: r.allowed, count: r.count, limit: body.limit }
    }

    @Get("compare")
    public async compare(
        @Query("key") key = "demo",
        @Query("limit") limit = "5",
        @Query("windowMs") windowMs = "2000",
        @Query("n") n = "10",
    ): Promise<Record<string, unknown>> {
        const result = await this.service.compare(key, Number(limit), Number(windowMs), Number(n))
        return { ...result }
    }
}
